### ISSUE TITLE
Discussion about TernGrad

### ISSUE THREAD

**Issue by nachtsky1077** [sands-lab/grace#14]
Thanks for the great work of integrating the gradient compression technique to horovod!

I'm also implementing a tensorflow version of TernGrad under horovod framework, and tried to minimize the effect of overhead (compress, decompress) by developing a horovod op and supporting cuda kernels. I've already implemented an allgathered version but it seems cost too much memory when scaling to large number of nodes. Hence, I'm trying to implement an allreduce version, which needs the "scaler sharing" technique mentioned in the original paper. But I met severe convergence issue on BERT pretraining task, comparing to the allgather version without scaler sharing I developed. I'm wondering have you guys tried out scaler sharing and allreduce for TernGrad?

Thanks!

**Comment by hangxu0304** [comment]
Hi,

We haven't tried the scaler sharing TernGrad. Even though it can reduce the communication overhead, the accuracy is actually compromised. And the actual performance may vary from model to model. I suggest you can try multiple global scaler, and use more bits in each worker to represent TernGrad compressed gradients.
We are also working on the allreduce for quantized gradients. Any comments/discussions are welcome.

Hang

**Comment by nachtsky1077** [comment]
By multiple global scaler, do you mean different scalers for different gradients/layers?

Also is there mailing threads or community about gradient compression? Thanks!

**Comment by hangxu0304** [comment]
My initial idea was to first allgather different scalers from all nodes (layer-wise), and then use this multiple scalers as the quantization states for local gradients. Of course, in order to make the quantized gradients in different nodes are addable, you will have to adjust the states values. Plus, you will need an efficient encoding to save more bits (e.g. put 6 quantized values into one 32-bits integer). 
Regarding the community, I'm afraid there is no such things right now. It will come out when this technique becomes practical in terms of performance. We have already addressed one of the challenges, i.e. [OmniReduce](https://sands.kaust.edu.sa/project/omnireduce/), an efficient sparse tensor Allreduce. Soon this feature will be available in GRACE. 

Hang

**Comment by nachtsky1077** [comment]
I've implemented the Allreduce version of TernGrad with scaler sharing. The pre-training results on BERT-base showed that it could converge but slower than the baseline and suffered from accuracy loss, but no substantial throughput increase was observed. I also tried on BERT-large, and saw 70% throughput increase comparing with the baseline (my test was on 16 V100 GPU on 2 machines).

In order to fix the convergence speed/rate issue, I tried out the vanilla error compensation mechanism, but the results is not as expected (it seems to slow down the convergence). So I'm wondering do you have any advice of applying the error feedback? (I did some search and found out that it might couple with the optimizer).

**Comment by hangxu0304** [comment]
In our [survey](https://repository.kaust.edu.sa/handle/10754/662495) paper, we observed that **Error Feedback(EF)**  always improves the accuracy of sparsification methods, but does not work well for some quantization methods, including QSGD, TernGrad and SignSGD. I'm not sure how much this issue is related to the optimizer, however, based on my own intuition, if the quantization method does not approximate the original values very well, by applying EF you will accumulate more and more errors and the training process will not be able to correct it. Below is our experiment result of training ResNet50 on ImageNet using TernGrad(no scaler sharing) with/without EF, which also validates your observations.
Some compressors also propose their own EF algorithm, such as PowerSGD, EFSignSGD, DGC. In EFSignSGD, they introduce [learning rate](https://github.com/sands-lab/grace/blob/2125adcc1905582a92cb363faa1ab2e412b21c2d/grace_dl/dist/memory/efsignsgd.py#L12) for error compensation. I would suggest you try this first.

![image](https://user-images.githubusercontent.com/52653060/101275831-d6f83400-37b9-11eb-91d3-bc844f75a3d0.png)

**Comment by nachtsky1077** [comment]
Thanks for your advice, I'll try out signSGD and sparsification methods later.

Meanwhile, as you've mentioned above that if the compressed gradients do not approximate the original gradients well, the error-feedback mechanism might amplify the negative effect brought in by compression, hence the accuracy and convergence will be harmed badly. I'm trying to figure out metrics to evaluate the original gradients and compressed gradients, like the sign difference and distribution difference, so as to better understand which layer/gradients are suitable for gradient compression (the author of TernGrad paper mentioned that they don't recommend apply gradient compression to the output layer). With the help of these statistics, I'm thinking of an adaptive method(layer-wise or even gradient-wise) of gradient compression.

Any advice regrading this idea, or have you tried out similar methods? Thanks!

**Comment by hangxu0304** [comment]
Generally, we don't recommend applying compression on rank one gradients in practice, since these are mostly biases and their sizes are relatively small. For weights, people often use [entropy](https://ieeexplore.ieee.org/abstract/document/8486500) to measure the importance of each layer. If embedding layers are inherently sparse, then sparsification might work better than quantization methods.
We tested TopK(0.01 ratio) in BERT pre-training, and it shows almost the same convergence rate (the loss curve) as no compression. We will further validate this result and also try TernGrad to see the slow convergence issue you mentioned.

Hang

**Comment by nachtsky1077** [comment]
Great! For the TopK BERT pre-training, did you experiment on BERT-base or BERT-large, as I'm also considering testing sparsification methods?

**Comment by hangxu0304** [comment]
@ChenYuHo Which BERT model did you use?

**Comment by ChenYuHo** [comment]
BERT-large as [here](https://github.com/NVIDIA/DeepLearningExamples/blob/99b1c898cead5603c945721162270c2fe077b4a2/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh#L187)

**Comment by nachtsky1077** [comment]
> BERT-large as [here](https://github.com/NVIDIA/DeepLearningExamples/blob/99b1c898cead5603c945721162270c2fe077b4a2/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh#L187)

Just to confirm, did you use the TopK compressor implemented in this repo, and did you use residual error feedback? Thanks!

**Comment by hangxu0304** [comment]
> > BERT-large as [here](https://github.com/NVIDIA/DeepLearningExamples/blob/99b1c898cead5603c945721162270c2fe077b4a2/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh#L187)
> 
> Just to confirm, did you use the TopK compressor implemented in this repo, and did you use residual error feedback? Thanks!

Yes, both memory and TopK compressor are from GRACE. However, we just found a bug in our experiment, need to rerun it.

**Comment by nachtsky1077** [comment]
> > > BERT-large as [here](https://github.com/NVIDIA/DeepLearningExamples/blob/99b1c898cead5603c945721162270c2fe077b4a2/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh#L187)
> > 
> > 
> > Just to confirm, did you use the TopK compressor implemented in this repo, and did you use residual error feedback? Thanks!
> 
> Yes, both memory and TopK compressor are from GRACE. However, we just found a bug in our experiment, need to rerun it.

Ok, I'll give it a try as well on my side.
